velox/dwio/common/ColumnVisitors.h

```diff
@@ -137,6 +137,98 @@ class ExtractToGenericHook {
   ValueHook* hook_;
 };
 
+class ExtractStringDictionaryToGenericHook {
+ public:
+  using HookType = ValueHook;
+  static constexpr bool kSkipNulls = false;
+
+  ExtractStringDictionaryToGenericHook(
+      ValueHook* hook,
+      RowSet rows,
+      RawScanState* state)
+      : hook_(hook), rows_(rows), state_(state) {}
+
+  bool acceptsNulls() const {
+    return hook_->acceptsNulls();
+  }
+
+  template <typename T>
+  void addNull(vector_size_t rowIndex) {
+    hook_->addNull(rowIndex);
+  }
+
+  template <typename V>
+  void addValue(vector_size_t rowIndex, V value) {
+    VELOX_UNREACHABLE();
+  }
+
+  ValueHook& hook() {
+    return *hook_;
+  }
+
+ private:
+  int32_t dictionarySize() const {
+    return state_->dictionary.numValues;
+  }
+
+  ValueHook* const hook_;
+  RowSet const rows_;
+  RawScanState* state_;
+};
+
+template <>
+void ExtractStringDictionaryToGenericHook::addValue<int32_t>(
+    vector_size_t rowIndex,
+    int32_t value);
+
+class ExtractToHookWrapper {
```

Comment:
Why do we need this?  The decision should be made when we create the extract object, not per row

Reply:
In Parquet, the encoding of the string field is known in `PageReader::prepareDataPageV1/PageReader::prepareDataPageV2`, and the extract object is created before that. So when creating `dwio::common::StringColumnReadWithVisitorHelper` in `StringColumnReader::read`, I don’t know whether to choose `ExtractToGenericHook ` or `ExtractStringDictionaryToGenericHook`. In order to be compatible with DWRF/ORC, `ExtractToGenericHook` was selected. After calling `PageReader::prepareDataPageV1/PageReader::prepareDataPageV2`, Parquet will initialize `ExtractStringDictionaryToGenericHook` in `StringDictionaryColumnVisitor<TFilter, ExtractValues, isDense> toStringDictionaryColumnVisitor()`. If there is no `ExtractToHookWrapper`, the compilation will fail. This method is really not good. I am still thinking of other ways. Do you have any suggestions?

Reply:
Since it is only needed by parquet to make decision at page level, let's move it into string column reader in parquet.  Other formats should not pay for the row-wise decision here.

Reply:
Can you move this class and your copy of `ExtractStringDictionaryToGenericHook` to parquet `StringColumnReader.cpp` since it's only used there?

Reply:
I have moved it into `StringColumnReader.cpp`.